File: backend/fastapi_app/types/web3_types.py
from enum import Enum
from typing import Optional
from pydantic import BaseModel, Field

class Web3Network(Enum):
    MAINNET = "solana-mainnet"
    TESTNET = "solana-testnet"
    DEVNET = "solana-devnet"
    
class Web3ActionType(str,Enum):
    TRANSFER = "transfer"
    MINT = "mint"
    BURN = "burn"
    STAKE = "stake"
    UNSTAKE = "unstake"
    
class TransferRequest(BaseModel):
    network: Web3Network
    to_address: str
    amount: float
    token_mint: Optional[str] = None
    private_key: str = Field(..., exclude=True )
    
class Web3Response(BaseModel):
    task_id: str
    status: str
    message: str
    
class MintRequest(BaseModel):
    network: Web3Network
    token_mint: str
    destination: str
    amount: float
    private_key: str = Field(..., exclude=True)
    mint_authority: Optional[str] = None
    
class BurnRequest(BaseModel):
    network: Web3Network
    token_mint: str
    amount: float
    source: str
    private_key: str = Field(..., exclude=True)
    mint_authority: Optional[str] = None
    
class StakeRequest(BaseModel):
    network: Web3Network
    validator: str
    amount: float
    private_key: str = Field(..., exclude=True)
    
class UnstakeRequest(BaseModel):
    network: Web3Network
    stake_account: str
    private_key: str = Field(..., exclude=True)
    
class TransactionStatus(str, Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"

class TokenConfig(BaseModel):
    mint: str
    decimals: int

class GasConfig(BaseModel):
    priority_fee: Optional[int] = None
    max_fee: Optional[int] = None

class Web3Config(BaseModel):
    network: Web3Network
    action_type: Web3ActionType
    amount: Optional[str]
    recipient: Optional[str]
    token: Optional[TokenConfig]
    contract_address: Optional[str]
    function_name: Optional[str]
    parameters: Optional[Dict[str, Any]]
    gas_config: Optional[GasConfig]

File: frontend/src/types/web3.types.ts
export enum Web3Network {
    MAINNET = "solana-mainnet",
    TESTNET = "solana-testnet",
    DEVNET = "solana-devnet"
}

export enum Web3ActionType {
    TRANSFER = "transfer",
    MINT = "mint",
    BURN = "burn",
    STAKE = "stake",
    UNSTAKE = "unstake"
}

export enum TransactionStatus {
    PENDING = "pending",
    PROCESSING = "processing",
    COMPLETED = "completed",
    FAILED = "failed"
}

export interface TransferRequest {
    network: Web3Network;
    to_address: string;
    amount: number;
    token_mint?: string;
    private_key: string;
}

export interface MintRequest {
    network: Web3Network;
    token_mint: string;
    destination: string;
    amount: number;
    private_key: string;
    mint_authority?: string;
}

export interface BurnRequest {
    network: Web3Network;
    token_mint: string;
    amount: number;
    source: string;
    private_key: string;
    mint_authority?: string;
}

export interface StakeRequest {
    network: Web3Network;
    validator: string;
    amount: number;
    private_key: string;
}

export interface UnstakeRequest {
    network: Web3Network;
    stake_account: string;
    private_key: string;
}

export interface Web3Response {
    task_id: string;
    status: string;
    message: string;
}

export interface TokenConfig {
    mint: string;
    decimals: number;
}

export interface GasConfig {
    priority_fee?: number;
    max_fee?: number;
}

export interface Web3Config {
    network: Web3Network;
    action_type: Web3ActionType;
    amount?: string;
    recipient?: string;
    token?: TokenConfig;
    contract_address?: string;
    function_name?: string;
    parameters?: Record<string, any>;
    gas_config?: GasConfig;
}

export interface Web3Task {
    workflow_id: string;
    action_type: Web3ActionType;
    network: Web3Network;
    params: Record<string, any>;
    user_id: string;
    status: TransactionStatus;
    created_at: string;
    result: any;
    error?: string;
}

export interface TokenInfo {
    decimals: number;
    supply: number;
    freeze_authority?: string;
    mint_authority?: string;
}

export interface ValidatorInfo {
    pubkey: string;
    node_pubkey: string;
    activated_stake: number;
    commission: number;
    last_vote: number;
    root_slot: number;
    credits: number;
    epoch_vote_account: boolean;
}